src/EditorFeatures/Core/FindUsages/AggregateFindUsagesContext.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindUsages;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.FindUsages
+{
+    /// <summary>
+    /// Helper class that allows us to provide a unified Find-Usages progress over multiple
+    /// subroutines computing different types of usages.
+    /// </summary>
+    internal class FindUsagesContextAggregator
+    {
+        /// <summary>
+        /// Real context object we'll forward messages to.
+        /// </summary>
+        private readonly IFindUsagesContext _context;
+
+        /// <summary>
+        /// Underlying context objects that we'll receive messages from and forward to <see cref="_context"/>
+        /// </summary>
+        private readonly List<ForwardingFindUsagesContext> _underlyingContexts = new List<ForwardingFindUsagesContext>();
+
+        private readonly object _gate = new object();
+        private readonly Dictionary<ForwardingFindUsagesContext, (int current, int maxium)> _underlyingContextToProgress =
```

Comment:
Rather than keeping both a Dictionary with the contexts as keys and also a List, just have the single Dictionary and you just enumerate the keys?

Reply:
Done!